### PR TITLE
Blog app: add authorization rules

### PR DESCRIPTION
**Pull request by noelfoka**
## [Blog app: add authorization rules](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/blog_app_add_authorization.md#blog-app-add-authorization-rules)

In this branch, We:

- [ ] Install CanCanCan in my project.
- [ ] Add a role column to the users table. 
- [ ] A user can delete a post if it is theirs or if they have an admin role (column role has value "admin").
- [ ] A user can delete a comment if it is theirs or if they have an admin role (column role has value "admin").

### CODE REVIEW

File: app/controllers/comments_controller.rb [hunk 1]
```diff
@@ -9,7 +10,7 @@ def new
   def create
     @user = User.find(params[:user_id])
     @post = Post.find(params[:post_id])
-    @comment = @post.comments.new(comment_params.merge(author: @user))
+    @comment = @post.comments.new(comment_params.merge(user: @user))
```
Comment (DammyShittu) [inline app/controllers/comments_controller.rb]: - [ ] I cannot create a comment to test the comments authorization. Kindly resolve the error.

  image

File: app/models/user.rb [hunk 2]
```diff
@@ -11,6 +11,7 @@ class User < ApplicationRecord
   attribute :bio, :text
   attribute :post_counter, :integer, default: 0
   attribute :photo, :string
+  enum role: %i[user admin]
```
Comment (DammyShittu) [inline app/models/user.rb:14]: - [ ] Kindly verify your app functionalities. In your migration file you added a `role` to users table with a column type of `string`, but you are declaring `enum` for it in the model file. In rails, `enum` is typically an integer column type, not a string. I would suggest you stick to the string type you defined. I could not make a user an admin until I removed this line.

File: app/models/ability.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
+class Ability
+  include CanCan::Ability
+
+  def initialize(user)
+    # Define abilities for the user here. For example:
+    user ||= User.new
+    if user.admin?
```
Comment (DammyShittu) [inline app/models/ability.rb]: - [ ] After removing that line in the model file, I ran into another error. You are calling an `admin?` method which you did not define anywhere in your User model. Kindly define the method.

  image

- [ ] Kindly go through your app again and ensure that only admins and users who authored a post and comment can delete them.

File: app/views/posts/show.html.erb [hunk 4]
```diff
@@ -11,18 +11,27 @@
           <span class="card-text">Likes: <%= @post.likes_counter%></span>
         </div>
       </div>
-      <h5>Post Title: <%= @post.title %> </h5>
+      <h5>Post Title: <%= @post.title %></h5>
       <p class="card-text"><%= @post.text %></p>
+      <% if @post.author.id == current_user.id || current_user.admin? %>
```
Comment (DammyShittu) [inline app/views/posts/show.html.erb:16]: - **[OPTIONAL]** It would be nice to use the `can?` method as that is part of the tools `cancancan` provides.

**Review by DammyShittu — CHANGES_REQUESTED** [review]
Hi @noelfoka and @rbd3,

## Highlights
✔️ Role column in users table

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by cynthiainga — APPROVED** [review]
Hi @noelfoka & @rbd3, 

Your project is complete! There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 



## Highlights

- [ ] All requested changes are made ✔️ 
- [ ] No linters errors ✔️ 
- [ ] Role column in users table ✔️ 
- [ ] Correct Gitflow ✔️ 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were missed by the previous reviewer and addressing them will really improve your application._

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**